scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java

```diff
@@ -0,0 +1,184 @@
+                                  Pair.of(activityToSimulate.start, activityToSimulate.activity));
+      simulateSchedule(schedule);
+      activitiesInserted.add(activityToSimulate);
+    }
+  }
+
+
+  /**
+   * Get the simulation results from the Duration.ZERO to the current simulation time point
+   * @return the simulation results
+   */
+  public SimulationResults getSimulationResults(){
+    return getSimulationResultsUntil(curTime);
+  }
+
+  /**
+   * Get the simulation results from the Duration.ZERO to a specified end time point.
+   * The provided simulation results might cover more than the required time period.
+   * @return the simulation results
+   */
+  public SimulationResults getSimulationResultsUntil(Duration endTime){
+    //if previous results cover a bigger period, we return do not regenerate
+    if(lastSimResults == null || endTime.longerThan(lastSimResultsEnd)) {
+      Map<String, ActivityInstanceId> convertedTaskToPlannedDir = new HashMap<>();
+      taskToPlannedDirective.forEach((taskId, plannedDirective)->
+                                         convertedTaskToPlannedDir.put(taskId.id(), plannedDirective));
+      lastSimResults = engine.computeResults(
+          engine,
+          Instant.now(),
+          endTime,
+          convertedTaskToPlannedDir,
+          timeline,
+          missionModel);
+      lastSimResultsEnd = endTime;
+      //while sim results may not be up to date with curTime, a regeneration has taken place after the last insertion
+    }
+    return lastSimResults;
+  }
+
+  private void simulateSchedule(Map<ActivityInstanceId, Pair<Duration, SerializedActivity>> schedule){
+
+    for (final var entry : schedule.entrySet()) {
+      final var directiveId = entry.getKey();
+      final var startOffset = entry.getValue().getLeft();
+      final var directive = entry.getValue().getRight();
+
+      final var taskId = engine.initiateTaskFromInput(missionModel, directive);
+      engine.scheduleTask(taskId, startOffset);
+      plannedDirectiveToTask.put(directiveId,taskId);
+      taskToPlannedDirective.put(taskId, directiveId);
+    }
+
+    while (true) {
+      final var batch = engine.extractNextJobs(Duration.MAX_VALUE);
+      // Increment real time, if necessary.
+      final var delta = batch.offsetFromStart().minus(curTime);
+      curTime = batch.offsetFromStart();
+      timeline.add(delta);
+      // TODO: Advance a dense time counter so that future tasks are strictly ordered relative to these,
+      //   even if they occur at the same real time.
+
+      // Run the jobs in this batch.
+      final var commit = engine.performJobs(batch.jobs(), cells, curTime, Duration.MAX_VALUE, missionModel);
+      timeline.add(commit);
+
+      var exit = true;
+      for(var taskId : taskToPlannedDirective.keySet()){
+        if(!engine.isTaskComplete(taskId)){
+          exit = false;
+          break;
+        }
+      }
+      if(exit){
+        break;
+      }
```

Comment:
Another stylistic change but I'd probably write this as:
```java
      // Exit IFF all tasks are complete
      if (taskToPlannedDirective.keySet().stream().allMatch(taskId -> engine.isTaskComplete(taskId))) {
        break;
      }
```

Reply:
beautiful, I'll take it.